### DISCUSSION TITLE
CUDA gradient execution with respect to all inputs (*double)

### DISCUSSION THREAD

**Discussion by stebos100** [EnzymeAD/Enzyme#1780] [General]
Hi there everyone ! 

I was wondering if the following is possible,

lets say we have a function of the following format: 

`
  double generateGbmMatricesPtr ( double* Ints, double* instantaneousInts, double* vols, , int cols ) {

        double* gbms;
        double finalVal = 0;

        for (int j = 1; j < cols; j++)
        {
            const T drift = (Ints[j] - instantaneousInts[j]) - 0.5 * vols[j] * vols[j];

            const T diffusion = vols[j] * sqrtDt * randomNumbers[j];

            gbms[j] = gbms[j - 1] * exp(drift * dt + diffusion);

        }

        for (int j = 0; j < cols; j++)
        {
            finalVal += gbms[j];

        }

        return finalVal;
    }

`

Is there a way that Enzyme can differentiate the function with respect to all input vairables labelled 

`  double* Ints, double* instantaneousInts, double* vols,`

i.e there will be 120 inputs for all variables ? and we want to store the derivative values of those variables in a return double* array  using the reverse mode ? I assume that the Forward mode would be executed thread wise with respect to one input variable, but in this instance we want to differentiate the function w.r.t all 120 Interest rates and volatilities in one go.  

Looking forward to hearing from you !

**Comment by tgymnich** [comment]
I am missing some context, but this should be a good starting point: https://fwd.gymni.ch/YcMwSq

**Reply by stebos100** [reply]
Hi @tgymnich, 

thanks for getting back to me ! I just have one final question, if I wanted to do this in the CUDA context, how would I apply this ?

**Reply by tgymnich** [reply]
Here is our CUDA example code: https://fwd.gymni.ch/yRVp4s you can use to adapt the previous example.
https://enzyme.mit.edu/getting_started/CUDAGuide/
